network_simulator/network_topology.py

```diff
@@ -0,0 +1,278 @@
+            else:
+                device_addrs.add(host.address)
+                host_addrs.add(host.address)
+
+        # Check that set(end_1_addr, end_2_addr) is always unique, regardless
+        # of order
+        link_ends = set()
+        for link in self.links:
+            if link.end_1_addr is None or link.end_2_addr is None \
+                    or link.static_delay_sec is None \
+                    or link.capacity_bps is None \
+                    or link.buffer is None \
+                    or link.buffer.buffer_size_bits is None:
+                raise ValueError("Link " + link + " did not have required "
+                                 "properties set")
+            if link.end_1_device is not None or link.end_2_device is not None:
+                raise ValueError("Ends of Link " + link + " had Devices set")
+            if link.buffer.queue is not None:
+                raise ValueError("Link buffer " + link.buffer + " of Link "
+                                 + link + "had non-required properties set")
+
+            if {link.end_1_addr, link.end_2_addr} in link_ends:
+                raise ValueError("Link " + link + " had same ends as another "
+                                 "Link.")
+            else:
+                link_ends.add({link.end_1_addr, link.end_2_addr})
+
+        # Check for unique Flow IDs
+        flow_ids = set()
+        for flow in self.flows:
+            if flow.source_addr is None or flow.dest_addr is None or \
+                    flow.flow_id is None or flow.data_size_bits is None or \
+                    flow.start_time is None:
+                raise ValueError("Flow " + flow + " did not have required "
+                                 "properties")
+            if flow.source is not None or flow.dest is not None:
+                raise ValueError("Flow " + flow + " had a source or dest "
+                                 "Device set.")
+            if flow.source_addr not in host_addrs or \
+                    flow.dest_addr not in host_addrs:
+                raise ValueError("Flow " + flow + " had non-Host source or "
+                                 "destination address")
+
+            if flow.flow_id in flow_ids:
+                raise ValueError("Flow " + flow + " had same ID as another "
+                                 "Flow.")
+            else:
+                flow_ids.add(flow.flow_id)
+
+    def __get_device_with_addr(self, device_addr):
+        devices = copy.copy(self.hosts)
+        devices.append(self.routers)
+        for device in devices:
+            if device.address == device_addr:
+                return device
+
+        return None
+
+    def __get_host_with_addr(self, host_addr):
+        # Host address should be unique among Devices
+        host = self.__get_device_with_addr(host_addr)
+        if host is not None and not isinstance(host, Host):
+            raise ValueError("Tried to get Host with address " + host_addr +
```

Comment:
Okay, I don't think we need error.py for now then. We can add it later if we want to make the errors change names but it's not a big deal.